cpp/ql/lib/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll

```diff
@@ -1549,7 +1549,8 @@ private float getGuardedUpperBound(VariableAccess guardedAccess) {
     // that there is one predecessor, albeit somewhat conservative.
     exists(unique(BasicBlock b | b = def.(BasicBlock).getAPredecessor())) and
     guardedAccess = def.getAUse(v) and
-    result = max(float ub | upperBoundFromGuard(guard, guardVa, ub, branch))
+    result = max(float ub | upperBoundFromGuard(guard, guardVa, ub, branch)) and
+    not exists(Expr e | e = guard.getAChild+() | convertedExprMightOverflow(e))
```

Comment:
This predicate looks at any child of `guard` including expressions that does not affect `guardVa`.

I was considering extending this to: 
`not exists(Expr e | e = guard.getAChild+() and e = guardVa.getParent*() | convertedExprMightOverflow(e))`

but I was unable to construct an expression where this had any effect.

Reply:
It's a bit scary that we have to exclude overflows in any child of the guard. This is not something we've had to do before, I think. I would have hoped that we could fix it inside `upperBoundFromGuard` somehow. Is the real problem that we are overflowing in `linearAccessImpl`?

Reply:
I have the same concerns as @MathiasVP. I'd also like to think a bit more about the root cause, but at least we have this PR if we don't find a better solution before Monday.

Reply:
I agree there might be a better solution for this, but I'm unsure on where we should add this.

The predicate `getTruncatedUpperBounds` does handle the overflows correctly, but the addition of predicate `getGuardedUpperBound` was add specifically to handle cases where `getTruncatedUpperBounds` would widen the result inside a loop and hence miss obvious cases.